neuro_evolution_ctrnn/brains/CNN_CTRNN.py

```diff
@@ -19,33 +19,29 @@ def __init__(self, input_space: Space, output_space: Space, individual: np.ndarr
         self.config = config
-        cnn_size, ctrnn_size, cnn_output_space = self._get_sub_individual_size(config, input_space, output_space)
+        metadata_dict = self.get_free_parameter_metadata(config, input_space, output_space)
+        cnn_output_space = metadata_dict['_cnn_output_space']
+        cnn_size = Cnn.get_individual_size(config.cnn_conf, input_space, cnn_output_space)
+        ctrnn_size = ContinuousTimeRNN.get_individual_size(config.ctrnn_conf, cnn_output_space, output_space)
         ind_cnn = individual[0:cnn_size]
         ind_ctrnn = individual[cnn_size:cnn_size + ctrnn_size]
-        self.cnn = Cnn(input_space=input_space, output_space=cnn_output_space, config=config.cnn_conf,
-                       individual=ind_cnn)
+        self.cnn = Cnn(input_space=input_space, output_space=cnn_output_space,
+                       config=config.cnn_conf, individual=ind_cnn)
         self.ctrnn = ContinuousTimeRNN(input_space=cnn_output_space, output_space=output_space,
                                        config=config.ctrnn_conf, individual=ind_ctrnn)
 
     def step(self, ob: np.ndarray) -> Union[np.ndarray, np.generic]:
-        x = torch.from_numpy(np.array([ob], dtype=np.float32)).permute(0, 3, 1, 2)
-        cnn_out = self.cnn.forward(x=x)
-        return self.ctrnn.step(ob=cnn_out.numpy())
+        return self.ctrnn.step(self.cnn.step(ob))
 
     @classmethod
-    def _get_sub_individual_size(cls, config, input_space, output_space):
+    def get_free_parameter_metadata(cls, config: CnnCtrnnCfg, input_space: Space, output_space: Space):
         cnn_output_space = Cnn.get_output_shape(config=config.cnn_conf, input_space=input_space)
-
-        cnn_size = Cnn.get_individual_size(config=config.cnn_conf, input_space=input_space,
-                                           output_space=cnn_output_space)
-        ctrnn_size = ContinuousTimeRNN.get_individual_size(config=config.ctrnn_conf, input_space=cnn_output_space,
-                                                           output_space=output_space)
-        return cnn_size, ctrnn_size, cnn_output_space
-
-    @classmethod
-    def get_individual_size(cls, config: CnnCtrnnCfg, input_space: Space, output_space: Space):
-        a, b, _ = cls._get_sub_individual_size(config, input_space, output_space)
-        return a + b
+        cnn_metadata = Cnn.get_free_parameter_metadata(config=config.cnn_conf, input_space=input_space,
+                                                       output_space=cnn_output_space)
+        ctrnn_metadata = ContinuousTimeRNN.get_free_parameter_metadata(config=config.ctrnn_conf,
+                                                                       input_space=cnn_output_space,
+                                                                       output_space=output_space)
+        return {'CNN': cnn_metadata, 'CTRNN': ctrnn_metadata, '_cnn_output_space': cnn_output_space}
```

Comment:
I would place the strings `_cnn_output_space`, `CNN` and `CTRNN` into a dict at the top of this file and reference the strings here with the key from the dict. So that in the constructor you can also just use the key.

Reply:
I added constant `_usage_key_cnn_output = '_cnn_output_space'` to the class